ray-curator/ray_curator/stages/deduplication/semantic/kmeans.py

```diff
@@ -0,0 +1,371 @@
+        verbose (bool): Whether to print verbose output.
+        embedding_dim (int | None): The dimension of the embedding. This helps us read data into smaller chunks.
+        input_filetype (Literal["jsonl", "parquet"]): The type of the input file
+        input_file_extensions (list[str] | None): The file extensions of the input files. If not provided, we will infer it from the filetype.
+        read_kwargs (dict[dict]): Keyword arguments for the read stage.
+        write_kwargs (dict[dict]): Keyword arguments for the write stage.
+        max_iter (int): The maximum number of iterations to run.
+        tol (float): Tolerance for stopping criteria of the kmeans algorithm.
+        random_state (int): Seed for the random number generator. Unseeded by default. Does not currently fully guarantee the exact same results.
+        init (Literal["k-means||", "random"] | np.ndarray): 'scalable-k-means++' or 'k-means||': Uses fast and stable scalable kmeans++ initialization. 'random': Choose 'n_cluster' observations (rows) at random from data for the initial centroids. If an ndarray is passed, it should be of shape (n_clusters, n_features) and gives the initial centers.
+        n_init (int | Literal["auto"]): Number of times the k-means algorithm will be run with different centroid seeds. The final results will be the best output of n_init consecutive runs in terms of inertia.
+        oversampling_factor (float): The amount of points to sample in scalable k-means++ initialization for potential centroids. Increasing this value can lead to better initial centroids at the cost of memory. The total number of centroids sampled in scalable k-means++ is oversampling_factor * n_clusters * 8.
+        max_samples_per_batch (int): The number of data samples to use for batches of the pairwise distance computation. This computation is done throughout both fit predict. The default should suit most cases. The total number of elements in the batched pairwise distance computation is max_samples_per_batch * n_clusters. It might become necessary to lower this number when n_clusters becomes prohibitively large.
+    """
+
+    def __post_init__(self):
+        """Initialize parent class after dataclass initialization."""
+        super().__init__()
+
+    def decompose(self) -> list[ProcessingStage]:
+        # Set default file extensions based on input_filetype if not provided
+        file_extensions = FILETYPE_TO_EXTENSIONS.get(self.input_filetype, [])
+        if not file_extensions:
+            msg = f"Unsupported filetype: {self.input_filetype}"
+            raise ValueError(msg)
+
+        return [
+            FilePartitioningStage(
+                file_paths=self.input_path,
+                file_extensions=file_extensions,
+                files_per_partition=1,  # We set this to one, and then the RaftActor will break it up into smaller groups
+            ),
```

Comment:
I wonder if we should pass storage options here in FilePartitioningStage. 

I passed it here: 

https://github.com/NVIDIA-NeMo/Curator/pull/918/files#diff-6ec921fa8f277a57d283334f28937023416eb1c0318c72899fc6490f5a3a3989R172

Reply:
Just noting that this will fail today (which is fine, we can merge my PR after this). so we are good.